Fix empty vault token due to race condition

kreisys opened the pull request:
A `while` loop wrapped in an ssh call never gets invoked if the ssh call is run before the instance is accepting ssh connections. In that case we would get a `connection refused` and proceed to the next line in the bootstrapper provisioner script due to no `set -e`.

johnalotoski commented on modules/terraform.nix:
```diff
@@ -5,6 +5,8 @@ let
 
   kms2region = kms: if kms == null then null else builtins.elemAt (lib.splitString ":" kms) 3;
 
+  relEncryptedFolder = lib.last (builtins.split "-" (toString config.secrets.encryptedRoot));
```
Suggest making relEncryptedFolder prem infraType friendly here also.  This older def breaks the prem case due to both the split match and the config secrets reference.  Example [here](https://github.com/input-output-hk/bitte/blob/master/modules/terraform/tf-options.nix#L34-L39).

kreisys replied:
I see this logic repeated all over the place. I think it makes more sense to move your version into a single source of truth and use that throughout the codebase. But that would make more sense in its own PR (that I would be happy to create).

johnalotoski replied:
Here is another example to make it clearer: [link](https://github.com/input-output-hk/bitte/pull/105/files#diff-1582854d2aa689ad355e6623148d243c202547dfeb78c8a708621314e427a6dbR9-R11).  This logic diff which is scattered should be disappearing in the future as the prem and aws encryption re-converge on the same module usage: sops instead of age for prem case most likely.  So it's up to you whether you'd prefer to do a faster couple line change now to maintain compat, or invest more time into cleaning something up that will be disappearing in the near future anyway.

kreisys commented:
@johnalotoski I tested it, there was a silly mistake after all. Anyways I fixed it (`config.infraType` -> `config.cluster.infraType`) and squashed so that nobody will ever know 🙃
@manveru can I have approval #2?

manveru commented on modules/terraform.nix:
```diff
@@ -175,20 +165,42 @@ let
+    if ! test -f ${relEncryptedFolder}/nomad.bootstrap.enc.json; then
+      echo
+      echo Waiting for bootstrapping on core-1 to finish for nomad /var/lib/nomad/bootstrap.token ...
+      while ! ${ssh} ${sshTarget} -- test -f /var/lib/nomad/bootstrap.token &>/dev/null; do
```
`test -s` here please :)

manveru commented on modules/terraform.nix:
```diff
@@ -928,7 +943,18 @@ let
+        };
+
+        bootstrapperCommand = lib.mkOption {
+          type = with lib.types; nullOr (functionTo str);
```
we have `with lib.types` above already. I'd prefer `inherit` instead, since it's faster and easier to understand (and gives better error messages).

kreisys commented on modules/terraform.nix:
```diff
@@ -5,6 +5,10 @@ let
   kms2region = kms: if kms == null then null else builtins.elemAt (lib.splitString ":" kms) 3;
 
+  relEncryptedFolder = let
+    encPathStr = if (config.cluster.infraType == "aws") then (toString config.secrets.encryptedRoot) else (toString config.age.encryptedRoot);
```
This logic is duplicated all over the place (and now with different variations as well). I'd very much like to factor it out into a single place but I'm not sure what would be the "correct" place for it.
